WIP: Instrumenting test to see if it runs differently in CI

rcritten opened the pull request:
Apparently a hidden service becomes unhidden on restoration
but I can't reproduce it manually. Run with some instrumentation
to see what is going on in CI.

Next run will be against latest 389-ds.

rcritten commented:
DNS failure, re-running.

rcritten commented:
After initial replication a restart of 389-ds fails with:

[28/Jan/2022:17:23:34.322678384 +0000] - ERR - dse_read_one_file - The entry cn=schema in file /etc/dirsrv/slapd-IPA-TEST/schema/99user.ldif (lineno: 1) is invalid, error code 21 (Invalid syntax) - attribute type ( 2.16.840.1.113730.3.8.11.61 NAME 'ipaWrappingKey' DESC 'PKCS 115.121.1.15 SINGLE-VALUE X-ORIGIN ( 'IPA v4.1' 'user defined' ) ): Failed to parse attribute, error(2 - Unexpected token) at ( v4.1' 'user defined' ) ))
[28/Jan/2022:17:23:34.324556148 +0000] - ERR - setup_internal_backends - Please edit the file to correct the reported problems and then restart the server.

In the schema (added in 2014) this is:

attributeTypes: (2.16.840.1.113730.3.8.11.61 NAME 'ipaWrappingKey' DESC 'PKCS#11 URI of the wrapping key' EQUALITY caseExactMatch SYNTAX 1.3.6.1.4.1.1466.115.121.1.15 SINGLE-VALUE X-ORIGIN 'IPA v4.1' 

@mreynolds389 what do you think? This is running against 389ds copr repo.

abbra commented:
@rcritten I think @tbordaz was looking at this -- there is '#' in the description that is incorrectly parsed and Thierry fixed its handling already in  2.0.13-1.

mreynolds389 commented:
> @rcritten I think @tbordaz was looking at this -- there is '#' in the description that is incorrectly parsed and Thierry fixed its handling already in 2.0.13-1.

This looks to be duplicate of:  https://github.com/389ds/389-ds-base/issues/4903

This fix is in the latest copr build, and we have a test case for this exact schema definition that is failing for you.  I just verified the test case is still passing on master branch so I'm not sure what is going on in your case (old copr build?)

rcritten commented:
So basically this isn't pulling from copr at all as it should. It's using 389-ds-base-2.1.0-20211116.fc35.x86_64.

I don't know if I'm misusing this thing or not. I'll try being a bit more brute force since it isn't playing nice.

rcritten commented:
Ok this looks like a change in behavior, for the better, in 389-ds. Logs from the copr build http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/e4313510-80b3-11ec-9677-fa163e04db38 show that a hidden service remains hidden.

This is expected and correct. So a bug in older 389-ds appeasr to have restored deleted values.

As I reported in ticket <>, when I manually reproduce this I start with an LDIF that looks like:
```
# entry-id: 514
dn: cn=KDC,cn=replica.example.test,cn=masters,cn=ipa,cn=etc,dc=example,dc=test
modifyTimestamp;adcsn-61f0725d000000030003;vucsn-61f0725d000000030003: 2022012
 5215749Z
modifiersName;adcsn-61f0725d000000030002;vucsn-61f0725d000000030002: uid=admin
 ,cn=users,cn=accounts,dc=example,dc=test
objectClass;vucsn-61f06e96000000030000: nsContainer
objectClass;vucsn-61f06e96000000030000: ipaConfigObject
objectClass;vucsn-61f06e96000000030000: top
cn;vucsn-61f06e96000000030000;mdcsn-61f06e96000000030000: KDC
ipaConfigString;vucsn-61f06e96000000030000: startOrder 10
ipaConfigString;vucsn-61f06eb3000000030000: kdcProxyEnabled
ipaConfigString;vucsn-61f06fb3000000030000: pkinitEnabled
ipaConfigString;vucsn-61f0725d000000030001: hiddenService
ipaConfigString;vucsn-61f06e96000000030000;vdcsn-61f06fef000000030000;deleted:
 configuredService
ipaConfigString;vucsn-61f06fef000000030001;vdcsn-61f0725d000000030000;deleted:
 enabledService
creatorsName;vucsn-61f06e96000000030000: cn=Directory Manager
createTimestamp;vucsn-61f06e96000000030000: 20220125214142Z
nsUniqueId: 8dd4ca9b-7e2711ec-9dd58ace-11c1cbb2
```

Both the deleted attributes appear after restoration in older 389-ds. The copr version does not restore this.

IMHO this is correct behavior. If we want to make the restored IPA server not hidden we should be assertive about it and not rely on the behavior of 389-ds.

@mreynolds389 does this ldif2db change look familiar?

flo-renaud commented:
> So basically this isn't pulling from copr at all as it should. It's using 389-ds-base-2.1.0-20211116.fc35.x86_64.
> 
> I don't know if I'm misusing this thing or not. I'll try being a bit more brute force since it isn't playing nice.

In order to use the 389ds copr repo, one needs to specify the vagrant image `name: freeipa/389ds-master-f35` (which you did) AND add `update_packages: True` in the `args:` section of the job definition (otherwise the 389-ds-base package doesn't get updated and the tests use the one from the image).

> IMHO this is correct behavior. If we want to make the restored IPA server not hidden we should be assertive about it and not rely on the behavior of 389-ds.

I would also expect `ipa-restore` to restore the replica as hidden. Initially the current behavior (restore as not hidden) was seen as a bug (https://pagure.io/freeipa/issue/7894) but the issue got closed with `This behavior is now accepted as part of the Hidden Replica feature`. If we revert this decision, we need to update the ticket and probably check the documentation.

tbordaz commented:
Regarding the deleted values I have not been able to reproduced (on F34 2.0.12)

<pre>
dn: cn=KDC,cn=replica.example.test,cn=masters,cn=ipa,cn=etc,dc=example,dc=test
...
ipaConfigString;vucsn-61f06e96000000030000;vdcsn-61f06fef000000030000;deleted:
 configuredService
ipaConfigString;vucsn-61f06fef000000030001;vdcsn-61f0725d000000030000;deleted:
 enabledService
</pre>

Those values are deleted. After import (ldif2db), do you get the 'configuredService' and 'enabledService' values ? What version are you testing ?

rcritten commented:
Unfortunately I didn't keep the log to a reproduction in this PR but I've reproduced it locally on a couple of F33 VMs I had lying around, they run 389-ds-base-1.4.4.17-1.fc33.x86_64. Maybe I'll backout the change to use copr and re-run to get the logs.

The reproduction of the server remaining hidden is at http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/e4313510-80b3-11ec-9677-fa163e04db38 and running against current 389-ds copr builds.

rcritten commented:
This test was run on F35 with 389-ds-base-2.0.12-1.fc35.x86_64.

It "passes" because the expectations are wrong.

If you look at http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/79b82142-7ee4-11ec-b103-fa163eda45c5/report.html and expand the test test_integration/test_replica_promotion.py::TestHiddenReplicaPromotion::test_hidden_replica_backup_and_restore you'll see the searches I executed just prior to the backup showing hidden and just after the restore showing the other states.

Pre-backup
```
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica1.cmd185:transport.py:557 ipaConfigString: startOrder 10
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica1.cmd185:transport.py:557 ipaConfigString: kdcProxyEnabled
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica1.cmd185:transport.py:557 ipaConfigString: pkinitEnabled
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica1.cmd185:transport.py:557 ipaConfigString: hiddenService
```
Post-restore
```
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica1.cmd219:transport.py:557 ipaConfigString: startOrder 10
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica1.cmd219:transport.py:557 ipaConfigString: configuredService
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica1.cmd219:transport.py:557 ipaConfigString: kdcProxyEnabled
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica1.cmd219:transport.py:557 ipaConfigString: pkinitEnabled
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica1.cmd219:transport.py:557 ipaConfigString: enabledService
DEBUG    ipatests.pytest_ipa.integration.host.Host.replica1.cmd219:transport.py:557 ipaConfigString: hiddenService
```

rcritten commented:
Bug filed against 389-ds, closing this PR.
